honeycombio/query_spec.go

```diff
@@ -0,0 +1,213 @@
+package honeycombio
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	honeycombio "github.com/kvrhdn/go-honeycombio"
+)
+
+var validQueryCalculationOps = []string{
+	"COUNT",
+	"SUM",
+	"AVG",
+	"COUNT_DISTINCT",
+	"MAX",
+	"MIN",
+	"P001",
+	"P01",
+	"P05",
+	"P10",
+	"P25",
+	"P50",
+	"P75",
+	"P90",
+	"P95",
+	"P99",
+	"P999",
+	"HEATMAP",
+}
+var validQueryFilterOps = []string{
+	"=",
+	"!=",
+	">",
+	">=",
+	"<",
+	"<=",
+	"starts-with",
+	"does-not-start-with",
+	"exists",
+	"does-not-exist",
+	"contains",
+	"does-not-contain",
+}
+
+type querySpecConstraints struct {
+	calculationCount int
+}
+
+func defaultQuerySpecConstraints() *querySpecConstraints {
+	return &querySpecConstraints{
+		calculationCount: 0,
+	}
+}
+
+func createQuerySpecSchema(constraints *querySpecConstraints) *schema.Schema {
+	if constraints == nil {
+		constraints = defaultQuerySpecConstraints()
+	}
+
+	return &schema.Schema{
+		Type:     schema.TypeSet,
+		Required: true,
+		MinItems: 1,
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"calculation": {
+					Type:     schema.TypeSet,
+					Required: true,
```

Comment:
Can probably make this optional:
>If no calculations are provided, COUNT is applied.

Reply:
You would think so, but Honeycomb actually refuses this: _request failed with status code 422: {"error":"trigger query requires exactly one calculation"}_

(You can test this by playing with `trigger_test.go` in `go-honeycombio`.)

Reply:
Hmm...
>a strict subset of a Query Specification. A Trigger Query Specification must contain exactly one item in calculations, and may contain breakdowns, filters, and/or filter_combination, and time_range, which defines the query’s duration. The HEATMAP calculation may not be used for the trigger calculation. time_range will default to frequency if frequency is specified, or to 900 seconds (15 min, the default for frequency) if frequency is not specified

So maybe it's required for triggers, optional for other Queries (boards)?

Reply:
I've split off the query block as a new data source `honeycombio_query` (which makes everything much simpler!) and I've now also made `calculation` optional.
Only drawback is that creating a trigger can fail if the query does not have exactly one calculation. But maybe we can add some validation for that later.

Reply:
Looks great!